hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractContainerCollector.java

```diff
@@ -0,0 +1,376 @@
+import com.hazelcast.spi.ManagedService;
+import com.hazelcast.spi.NodeEngine;
+import com.hazelcast.spi.SplitBrainHandlerService;
+import com.hazelcast.spi.SplitBrainMergePolicy;
+import com.hazelcast.spi.impl.PartitionSpecificRunnable;
+import com.hazelcast.spi.impl.operationexecutor.OperationExecutor;
+import com.hazelcast.spi.impl.operationservice.impl.OperationServiceImpl;
+import com.hazelcast.spi.merge.DiscardMergePolicy;
+import com.hazelcast.spi.merge.SplitBrainMergePolicyProvider;
+import com.hazelcast.spi.partition.IPartitionService;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.CountDownLatch;
+
+/**
+ * Collects non-threadsafe containers for {@link SplitBrainHandlerService}
+ * capable data structures.
+ * <p>
+ * The {@link com.hazelcast.internal.cluster.impl.ClusterMergeTask ClusterMergeTask}
+ * collects all mergeable data from all Hazelcast services. The problem is that it
+ * runs on an arbitrary thread, but the data is only modified by specific
+ * {@link com.hazelcast.spi.impl.operationexecutor.impl.PartitionOperationThread partition threads}.
+ * This can cause visibility issues due to a missing happens-before relation.
+ * <p>
+ * The collector fetches the data via {@link PartitionSpecificRunnable} (which
+ * run on the partition threads) and stores them in a {@link ConcurrentHashMap}.
+ * This guarantees the visibility for the {@code ClusterMergeTask}.
+ * <p>
+ * The collector can be implemented for partitioned and non-partitioned data
+ * structures via the base classes {@link AbstractPartitionedContainerCollector}
+ * and {@link AbstractNonPartitionedContainerCollector}.
+ * <p>
+ * <b>Note:</b> The collector will retain the collected containers for the merge runnable.
+ * So {@link ManagedService#reset()} is not allowed to clear or destroy the containers.
+ * To ensure this the collector removes the link between the data structure and a collected container via:
+ * <ul>
+ * <li>{@link Iterator#remove()} in {@link CollectPartitionedContainerRunnable#collectContainers()}
+ * for partitioned data structures</li>
+ * <li>{@link Map#remove(Object)} in {@link AbstractNonPartitionedContainerCollector#getContainer(String)}
+ * for non-partitioned data structures</li>
+ * </ul>
+ * The cleanup of the containers itself is done via {@link #destroy(Object)} after the merge is done
+ * or if the container is not collected. Containers from backup partitions are directly cleaned via
+ * {@link #destroyBackup(Object)}.
+ *
+ * @param <C> container of the data structure
+ */
+@SuppressWarnings("WeakerAccess")
+public abstract class AbstractContainerCollector<C> {
+
+    private final ConcurrentMap<Integer, Collection<C>> containersByPartitionId
```

Comment:
This `ConcurrentMap` ensures the visibility between the threads. It's populated by `PartitionSpecificRunnable` on the partition threads and read from by the `AbstractContainerMerger` on an arbitrary thread.